Run PhraseMatcher on Spans

peter-exos opened the pull request:
## Description
Addresses https://github.com/explosion/spaCy/issues/6839

Since you may be busy with the spacy 3 release (thank you and congrats!), I went ahead and tried to implement the requested feature myself. It was easier than I thought: 
I extended PhraseMatcher to run on Spans and Docs (like Matcher) by adding start and end index arguments to `PhraseMatcher.find_matches` and using them to constrain the loops. 
Created a regression test and a functional test. All tests passed.

I adjusted the relevant doc strings, but the website docs may need to be updated as well. I have not done that.

This is my first contribution to this repo. 
Please let me know if I am missing anything, or if I can improve something in this PR.

### Types of change
enhancement, matcher/feat

## Checklist
- [x] I have submitted the spaCy Contributor Agreement.
- [x] I ran the tests, and all new and existing tests passed.
- [ ] My changes don't require a change to the documentation, or if they do, I've added all required information.

svlandeg commented on spacy/tests/matcher/test_phrase_matcher.py:
```diff
@@ -323,3 +323,17 @@ def test_phrase_matcher_deprecated(en_vocab):
+    matches_doc = matcher(doc)
+    matches_span = matcher(span)
+    assert len(matches_doc) == 1
+    assert len(matches_span) == 1
```
Thanks for adding a few tests! I think these are great. Perhaps you could also add one that has multiple matches in the original `doc`, say 3 matches, and then you try and match against a `span` that would only include the middle one? Just to ensure that we're not matching (by accident) outside of the span's boundaries.

peter-exos replied:
Added another test (`test_span_v_doc_in_phrasematcher`).  
See comment below.

svlandeg replied:
Looks good!

svlandeg reviewed (approved):
Thanks for the PR! It's clear that you've done an effort to ensure consistency with the existing codebase, using the same `doclike` variable name, `E195` etc, that's really nice :-)

The solution itself is indeed quite elegant, I like it. I have one additional request for a unit test, but otherwise it looks good to me. Perhaps my colleagues want to add anything so let's leave this open for review just a bit longer.

svlandeg commented:
FYI - if you do want to update the docs already as well, you can find the relevant file under `website/docs/api/phrasematcher.md`. 
I think you'd mainly want to edit the `## PhraseMatcher.\_\_call\_\_` section.

Typically when we edit `.md` files, we apply the "prettier" tool for automatic reformatting (but it's not a big deal if you don't for small edits).

It's also no problem if you don't update the docs yet, then we'll just add it later, so no worries either way.

peter-exos commented:
Thank you for your great feedback!

I just added another test (`test_span_v_doc_in_phrasematcher`), where the pattern appears 3 times in the Doc, but only 1 time in the Span. Is this what you were thinking about? 
This should check that PhraseMatcher does not match the entire Doc when given a Span, but only the matches in the Span.
If you have any more suggestions how to make the tests more precise, please let me know.

I also updated the website. It's a small change, but please have a look.  
I did not install the website, so I could not see the actual rendering.

svlandeg commented:
This looks great to me, thanks!

adrianeboyd commented:
Very nice PR, thanks! I just made one small edit to the tokenizer.
